src/session/client-certificate.c

```diff
@@ -127,6 +132,97 @@ unsigned long long get_proc_pid_start_time (int dirfd)
   return start_time;
 }
 
+/* Fallback for get_ws_proc_fd() on older kernels which don't support enough pidfd API */
+static int
+get_ws_proc_fd_pid_time (void)
+{
+  struct ucred ucred;
+  socklen_t ucred_len = sizeof ucred;
+  if (getsockopt (STDIN_FILENO, SOL_SOCKET, SO_PEERCRED, &ucred, &ucred_len) != 0 ||
+      /* this is an inout parameter, be extra suspicious */
+      ucred_len < sizeof ucred)
+    {
+      debug ("failed to read stdin peer credentials: %m; not in socket mode?");
+      warnx ("Certificate authentication only supported with cockpit-session.socket");
+      exit_init_problem ("authentication-unavailable", "Certificate authentication only supported with cockpit-session.socket");
+    }
+
+  debug ("unix socket mode, ws peer pid %d", ucred.pid);
+  int ws_pid_dirfd = open_proc_pid (ucred.pid);
+  unsigned long long ws_start_time = get_proc_pid_start_time (ws_pid_dirfd);
+
+  int my_pid_dirfd = open_proc_pid (getpid ());
+  unsigned long long my_start_time = get_proc_pid_start_time (my_pid_dirfd);
+  close (my_pid_dirfd);
+
+  debug ("peer start time: %llu, my start time: %llu", ws_start_time, my_start_time);
+
+  /* Guard against pid recycling: If a malicious user captures ws, keeps the socket in a forked child and exits
+    * the original pid, they can trap a different user to login, get the old pid (pointing ot their cgroup), and
+    * capture their session. To prevent that, require that ws must have started earlier than ourselves. */
+  if (my_start_time < ws_start_time)
+    {
+      warnx ("start time of this process (%llu) is older than cockpit-ws (%llu), pid recycling attack?",
+              my_start_time, ws_start_time);
+      close (ws_pid_dirfd);
+      exit_init_problem ("access-denied", "implausible cockpit-ws start time");
+    }
+
+  return ws_pid_dirfd;
+}
+
+/* Get a /proc/[pid] dirfd for our Unix socket peer (i.e. cockpit-ws).
+ * We only support being called via cockpit-session.socket (i.e. Unix socket).
+ */
+static int
+get_ws_proc_fd (void)
+{
+#if defined(SO_PEERPIDFD) && defined(HAVE_PIDFD_GETPID)
+  int pidfd = -1;
+  socklen_t socklen = sizeof pidfd;
+  /* this is always the pidfd for the process that started the communication, it cannot be recycled */
+  if (getsockopt (STDIN_FILENO, SOL_SOCKET, SO_PEERPIDFD, &pidfd, &socklen) < 0)
+    {
+      if (errno == ENOPROTOOPT || errno == ENODATA)
+        {
+          debug ("SO_PEERPIDFD not supported: %m, falling back to pid/time check");
+          return get_ws_proc_fd_pid_time ();
+        }
+
+      warn ("Failed to get peer pidfd");
+      exit_init_problem ("access-denied", "Failed to get peer pidfd");
+    }
+  /* this is an inout parameter, be extra suspicious; this really Should Not Happen™, so bomb out */
+  if (socklen < sizeof pidfd)
+    errx (EX, "SO_PEERPIDFD returned too small result");
+
+  /* get pid for pidfd; from here on this is racy and could suffer from PID recycling */
+  pid_t pid = pidfd_getpid (pidfd);
```

Comment:
Can we get `ENOSYS` here if the glibc has that feature but the syscall is missing in the running kernel?

Reasons why this might be impossible:
 - the presence of `SO_PEERPIDFD` implies the existence of the `pidfd_getpid()` syscall; or
 - maybe it doesn't but we don't care, because none of our supported distributions fall into that particular crack...
   - either they don't ship the (theorized) kernel version with `SO_PEERPIDFD` but without `pidfd_getpid()`; or
   - they don't ship a glibc with `pidfd_getpid()` but without a capable kernel

Reply:
I was thinking about that as well when writing this. On the one hand it is a possible regression with weird/custom kernel systems, but I'm also highly nervous about tolerating failures here. So this is a judgement call, but to be honest I'd rather err on the side of "bad failure" than "bad acceptance".

Reply:
Kept the strictness for now, and added a comment.

Reply:
I agree with "be strict until we find out it's a problem".